common/scala/src/main/scala/whisk/common/ForcableSemaphore.scala

```diff
@@ -121,4 +139,11 @@ class ForcableSemaphore(maxAllowed: Int) {
 
   /** Returns the number of currently available permits. Possibly negative. */
   def availablePermits: Int = sync.permits
+
+  /** Set the new maxAllowed value. */
+  def setMaxAllowed(newMaxAllowed: Int): Unit = {
```

Comment:
Do we need to synchronize here? Especially since `forceChangeState` and resetting maxAllowed are non-atomic operations.

Can we reuse bits from here: http://blog.teamlazerbeez.com/2009/04/20/javas-semaphore-resizing/?

I guess your `forceChangeState` method is roughly the same as `reducePermits` in the original `Semaphore` impl. modulo you calculate an absolute number whereas they pass in a difference, so that should be fine.

Reply:
thanks! learned a lot